Plugin crashes when upating `snowflake_resource_monitor`

**Provider Version**

`v0.72.0`

**Terraform Version**

`1.6.2`

**Describe the bug**

When trying to add a new user to `snowflake_resource_monitor` it crashes.

**Expected behavior**

It should update the resource monitor with the additional user

**Code samples and commands**

The error occured when adding `User2` to `notify_users`

```
resource "snowflake_resource_monitor" "demo" {
  name         = snowflake_warehouse.demo.name
  credit_quota = 5

  frequency       = "DAILY"
  start_timestamp = "2023-11-01T17:00:00-07:00"
  end_timestamp   = "2024-10-31T17:00:00-07:00"

  notify_triggers           = [75]
  suspend_trigger           = 85
  suspend_immediate_trigger = 95

  notify_users = ["User1", "User2"]
  warehouses   = [snowflake_warehouse.demo.name]
}
```

**Additional context**

```
snowflake_resource_monitor.demo: Modifying... [id=DEMO_WH]
╷
│ Error: Request cancelled
│
│   with snowflake_resource_monitor.demo,
│   on demo.tf line 7, in resource "snowflake_resource_monitor" "demo":
│    7: resource "snowflake_resource_monitor" "demo" {
│
│ The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.
╵

Stack trace from the terraform-provider-snowflake_v0.72.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x8 pc=0x103669608]

goroutine 16 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.UpdateResourceMonitor(0x14000f38300, {0x103f6a760?, 0x14000b0e340?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/resource_monitor.go:346 +0x1448
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x103fa88c0?, {0x103fa88c0?, 0x14000f2e870?}, 0xd?, {0x103f6a760?, 0x14000b0e340?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:800 +0x134
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x1400061fb20, {0x103fa88c0, 0x14000f2e870}, 0x14000ef5040, 0x14000f38180, {0x103f6a760, 0x14000b0e340})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:919 +0x6b0
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x14000c92018, {0x103fa88c0?, 0x14000f2e750?}, 0x14000f14820)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/grpc_provider.go:1060 +0xb78
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x14000c9c000, {0x103fa88c0?, 0x14000f03d40?}, 0x14000edb0a0)
	github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:859 +0x3bc
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x103f1dca0?, 0x14000c9c000}, {0x103fa88c0, 0x14000f03d40}, 0x14000edb030, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:467 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000780d20, {0x103fb7e20, 0x14000c0cb60}, 0x14000e41560, 0x14000c8a180, 0x104b64c78, 0x0)
	google.golang.org/grpc@v1.57.0/server.go:1360 +0xcb4
google.golang.org/grpc.(*Server).handleStream(0x14000780d20, {0x103fb7e20, 0x14000c0cb60}, 0x14000e41560, 0x0)
	google.golang.org/grpc@v1.57.0/server.go:1737 +0x82c
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	google.golang.org/grpc@v1.57.0/server.go:982 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.57.0/server.go:980 +0x16c

Error: The terraform-provider-snowflake_v0.72.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

--- a/pkg/resources/dynamic_table.go
+++ b/pkg/resources/dynamic_table.go
@@ -65,6 +65,7 @@ var dynamicTableShema = map[string]*schema.Schema{
 	"query": {
 		Type:             schema.TypeString,
 		Required:         true,
+		ForceNew:         true,
 		Description:      "Specifies the query to use to populate the dynamic table.",
 		DiffSuppressFunc: DiffSuppressStatement,
 	},

--- a/pkg/resources/dynamic_table_acceptance_test.go
+++ b/pkg/resources/dynamic_table_acceptance_test.go
@@ -269,7 +269,7 @@ func TestAcc_DynamicTable_issue2276(t *testing.T) {
 				ConfigVariables: m(),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "name", dynamicTableName),
-					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", fmt.Sprintf("select \"id\" from \"%v\".\"%v\".\"%v\"", acc.TestDatabaseName, acc.TestSchemaName, tableName)),
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", query),
 				),
 			},
 			{
@@ -277,7 +277,7 @@ func TestAcc_DynamicTable_issue2276(t *testing.T) {
 				ConfigVariables: m2,
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "name", dynamicTableName),
-					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", fmt.Sprintf("select \"data\" from \"%v\".\"%v\".\"%v\"", acc.TestDatabaseName, acc.TestSchemaName, tableName)),
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", newQuery),
 				),
 			},
 		},

--- a/pkg/resources/dynamic_table_acceptance_test.go
+++ b/pkg/resources/dynamic_table_acceptance_test.go
@@ -236,6 +236,54 @@ func TestAcc_DynamicTable_issue2134(t *testing.T) {
 	})
 }
 
+// TestAcc_DynamicTable_issue2276 proves https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2276 issue.
+func TestAcc_DynamicTable_issue2276(t *testing.T) {
+	dynamicTableName := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
+	tableName := dynamicTableName + "_table"
+	query := fmt.Sprintf(`select "id" from "%v"."%v"."%v"`, acc.TestDatabaseName, acc.TestSchemaName, tableName)
+	newQuery := fmt.Sprintf(`select "data" from "%v"."%v"."%v"`, acc.TestDatabaseName, acc.TestSchemaName, tableName)
+	m := func() map[string]config.Variable {
+		return map[string]config.Variable{
+			"name":       config.StringVariable(dynamicTableName),
+			"database":   config.StringVariable(acc.TestDatabaseName),
+			"schema":     config.StringVariable(acc.TestSchemaName),
+			"warehouse":  config.StringVariable(acc.TestWarehouseName),
+			"query":      config.StringVariable(query),
+			"comment":    config.StringVariable("Terraform acceptance test for GH issue 2276"),
+			"table_name": config.StringVariable(tableName),
+		}
+	}
+	m2 := m()
+	m2["query"] = config.StringVariable(newQuery)
+
+	resource.Test(t, resource.TestCase{
+		ProtoV6ProviderFactories: acc.TestAccProtoV6ProviderFactories,
+		PreCheck:                 func() { acc.TestAccPreCheck(t) },
+		TerraformVersionChecks: []tfversion.TerraformVersionCheck{
+			tfversion.RequireAbove(tfversion.Version1_5_0),
+		},
+		CheckDestroy: testAccCheckDynamicTableDestroy,
+		Steps: []resource.TestStep{
+			{
+				ConfigDirectory: config.TestStepDirectory(),
+				ConfigVariables: m(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "name", dynamicTableName),
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", fmt.Sprintf("select \"id\" from \"%v\".\"%v\".\"%v\"", acc.TestDatabaseName, acc.TestSchemaName, tableName)),
+				),
+			},
+			{
+				ConfigDirectory: acc.ConfigurationSameAsStepN(1),
+				ConfigVariables: m2,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "name", dynamicTableName),
+					resource.TestCheckResourceAttr("snowflake_dynamic_table.dt", "query", fmt.Sprintf("select \"data\" from \"%v\".\"%v\".\"%v\"", acc.TestDatabaseName, acc.TestSchemaName, tableName)),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckDynamicTableDestroy(s *terraform.State) error {
 	db := acc.TestAccProvider.Meta().(*sql.DB)
 	client := sdk.NewClientFromDB(db)

--- a/pkg/resources/resource_monitor.go
+++ b/pkg/resources/resource_monitor.go
@@ -327,7 +327,9 @@ func UpdateResourceMonitor(d *schema.ResourceData, meta interface{}) error {
 		for _, name := range userNames {
 			users = append(users, sdk.NotifiedUser{Name: name})
 		}
-		opts.NotifyUsers.Users = users
+		opts.NotifyUsers = &sdk.NotifyUsers{
+			Users: users,
+		}
 	}
 
 	if d.HasChange("credit_quota") {

--- a/pkg/resources/resource_monitor_acceptance_test.go
+++ b/pkg/resources/resource_monitor_acceptance_test.go
@@ -4,12 +4,14 @@ import (
 	"encoding/json"
 	"fmt"
 	"os"
+	"regexp"
 	"strings"
 	"testing"
 
 	acc "github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/acceptance"
 	"github.com/hashicorp/terraform-plugin-testing/helper/acctest"
 	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/stretchr/testify/require"
 )
 
 func TestAcc_ResourceMonitor(t *testing.T) {
@@ -214,6 +216,35 @@ resource "snowflake_resource_monitor" "test" {
 	suspend_immediate_trigger = 95
 }
 `, accName)
+}
+
+// TestAcc_ResourceMonitor_issue2167 proves https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/2167 issue.
+// Second step is purposely error, because tests TestAcc_ResourceMonitorUpdateNotifyUsers and TestAcc_ResourceMonitorNotifyUsers are still skipped.
+// It can be fixed with them.
+func TestAcc_ResourceMonitor_issue2167(t *testing.T) {
+	name := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
+	configNoUsers, err := resourceMonitorNotifyUsersConfig(name, []string{})
+	require.NoError(t, err)
+	config, err := resourceMonitorNotifyUsersConfig(name, []string{"does not matter"})
+	require.NoError(t, err)
+
+	resource.Test(t, resource.TestCase{
+		Providers:    acc.TestAccProviders(),
+		PreCheck:     func() { acc.TestAccPreCheck(t) },
+		CheckDestroy: nil,
+		Steps: []resource.TestStep{
+			{
+				Config: configNoUsers,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("snowflake_resource_monitor.test", "name", name),
+				),
+			},
+			{
+				Config:      config,
+				ExpectError: regexp.MustCompile(`.*exactly one of AlterResourceMonitorOptions fields \[Set NotifyUsers Triggers] must be set.*`),
+			},
+		},
+	})
 }
 
 func TestAcc_ResourceMonitorNotifyUsers(t *testing.T) {

--- a/pkg/resources/testdata/TestAcc_DynamicTable_issue2276/1/test.tf
+++ b/pkg/resources/testdata/TestAcc_DynamicTable_issue2276/1/test.tf
@@ -1,0 +1,27 @@
+resource "snowflake_table" "t" {
+  database        = var.database
+  schema          = var.schema
+  name            = var.table_name
+  change_tracking = true
+  column {
+    name = "id"
+    type = "NUMBER(38,0)"
+  }
+  column {
+    name = "data"
+    type = "VARCHAR(16)"
+  }
+}
+
+resource "snowflake_dynamic_table" "dt" {
+  depends_on = [snowflake_table.t]
+  name       = var.name
+  database   = var.database
+  schema     = var.schema
+  target_lag {
+    maximum_duration = "2 minutes"
+  }
+  warehouse = var.warehouse
+  query     = var.query
+  comment   = var.comment
+}

--- a/pkg/resources/testdata/TestAcc_DynamicTable_issue2276/1/variables.tf
+++ b/pkg/resources/testdata/TestAcc_DynamicTable_issue2276/1/variables.tf
@@ -1,0 +1,27 @@
+variable "name" {
+  type = string
+}
+
+variable "database" {
+  type = string
+}
+
+variable "schema" {
+  type = string
+}
+
+variable "warehouse" {
+  type = string
+}
+
+variable "query" {
+  type = string
+}
+
+variable "comment" {
+  type = string
+}
+
+variable "table_name" {
+  type = string
+}

